torch/_inductor/triton_bundler.py

```diff
@@ -0,0 +1,188 @@
+
+    @staticmethod
+    def is_enabled() -> bool:
+        from torch._inductor import config
+
+        if (b := config.bundle_triton_into_fx_graph_cache) is not None:
+            return b
+
+        if not config.is_fbcode():
+            return False
+
+        return justknobs_check("pytorch/remote_cache:bundle_triton_into_fx_graph_cache")
+
+    @classmethod
+    def begin_compile(cls) -> None:
+        if not TritonBundler.is_enabled():
+            return
+        assert cls._entries is None
+        cls._entries = OrderedSet()
+
+    @classmethod
+    def put(cls, kernel_hash: str, compile_time_ns: int, device: int) -> None:
+        if (entries := cls._entries) is not None:
+            # CachingAutotuner.__init__ unconditionally sets TRITON_CACHE_DIR
+            triton_cache_dir = os.getenv("TRITON_CACHE_DIR")
+            assert triton_cache_dir is not None
+            entries.add(
+                TritonBundleEntry(
+                    kernel_hash, compile_time_ns, device, triton_cache_dir
+                )
+            )
+
+    @classmethod
+    def collect(
+        cls,
+    ) -> Tuple[List[TritonKernelArtifacts], Optional[TritonBundlerMetadata]]:
+        if not TritonBundler.is_enabled():
+            cls._entries = None
+            return [], None
+
+        with dynamo_timed(
+            key="TritonBundler::collect", phase_name="TritonBundler", fwd_only=False
+        ):
+            if (entries := cls._entries) is not None:
+                result: List[TritonKernelArtifacts] = []
+                total_compile_time_ns = 0
+                for entry in entries:
+                    artifacts: List[TritonKernelArtifact] = []
+                    path = os.path.join(entry.directory, entry.kernel_hash)
+                    if not os.path.exists(path):
+                        continue
+                    for filename in os.listdir(path):
+                        filepath = os.path.join(path, filename)
+                        try:
+                            assert os.path.isfile(filepath)
+                            with open(filepath, "rb") as file:
+                                artifacts.append(
+                                    TritonKernelArtifact(filename, file.read())
+                                )
+                            counters["inductor"]["triton_bundler_save_kernel"] += 1
+                        except Exception:
+                            log.debug("failed to collect triton kernel", exc_info=True)
+                    if artifacts:
+                        result.append(
+                            TritonKernelArtifacts(
+                                entry.kernel_hash,
+                                entry.compile_time_ns,
+                                entry.device,
+                                artifacts,
+                            )
+                        )
+                        total_compile_time_ns += entry.compile_time_ns
```

Comment:
This potentially overrepresents end-to-end compile time because when parallel compilation is enabled the true wall time is less than the sum of all the individual compile times of each artifact

Reply:
What would you recommend here? doing max based on whether parallel compile is enabled or not?

Reply:
But you're not going to have infinite parallelism right. So max isn't right either.

It also gets murkier because you're going to use this to adjust compiler collective right? So now here's another asymmetric situation to worry about: a bundled cache is generated when parallel compile is on, let's suppose perfect parallelism so 20 kernels x 10s = 10s total.  But someone else only partially cache hits in a setting where parallel compile is off, so now you have 20 x 10s = 200s on the other ranks. NCCL timeout!

For NCCL timeout, it seems safe to estimate the *upper bound* of how much compile time is potentially saved. So summing is the right quantity to use for NCCL timeout. But for "time saved" analysis, you would rather have something more realistic that accounts for parallel compile, or need lots of disclaimers on the data.

Reply:
This is actually worse. Remember how @masnesral and I were explaining that we compile kernels twice once in subprocess and once in mainprocess. SO this number is the second compilation, not first.

But this is actually fine for NCCL since we are not using this number to adjust, we are using compile_fx time (already committed) since that includes this timing end-to-end. With a partial cache hit, it does get muddier though, but we are arleady fuding the number to be higher anyway